feat: introduce experimental `unstable_getServerSession` API (#4116)
nextauthjs/next-auth: docs/docs/configuration/nextjs.md

# Next.js

## `getServerSession`

> > 🚧 This is experimental and may be removed in the future.

Reviewer: ```suggestion
> 🚧 This is experimental and may be removed or changed in the future.
```

Let's keep this hanging... :sweat_smile: Might stick around for a while. Might not...

> When calling from server side i.e. in API routes or in `getServerSideProps`, we recommend use this function instead of `getSession` to retrieve the `session` object. This method is especially useful when you are using NextAuth.js with a database. Otherwise, if you only want to get the token, see [`getToken`](tutorials/securing-pages-and-api-routes#using-gettoken).

Reviewer: ```suggestion
When calling from server-side i.e. in API routes or in `getServerSideProps`, we recommend using this function instead of `getSession` to retrieve the `session` object. This method is especially useful when you are using NextAuth.js with a database. Otherwise, if you only want to get the token, see [`getToken`](tutorials/securing-pages-and-api-routes#using-gettoken).
```

`getServerSession` requires passing the `authOptions` object. This is the same object you would pass to `NextAuth` when initializing NextAuth.js.

In `[...nextauth.js]`:
```js 
export const authOptions: NextAuthOptions = {
  // your configs
}

export default NextAuth(authOptions);
```

In `getServerSideProps`:
```js

> import authOptions from 'pages/api/[...nextauth]'

Reviewer: ```suggestion
import { authOptions } from 'pages/api/[...nextauth]'
```

export async function getServerSideProps(context) {
  const session = await getServerSession(context.req, context.res, authOptions)

  if (session) {
    // do something with the session
  }

  return {
    props: {
      session,
    },

>   }

Reviewer: Should we return a redirect if unauthenticated? https://nextjs.org/docs/api-reference/data-fetching/get-server-side-props#redirect

Author: Totally!

}
```
In API routes:
```js

> import authOptions from 'pages/api/[...nextauth]'

Reviewer: ```suggestion
import { authOptions } from 'pages/api/[...nextauth]'
```

export async function handler(req, res) {
  const session = await getServerSession(req, res, authOptions)

  if (session) {

>     // do something with the session

Reviewer: should we rather have a different response when there is or there isn't a session? Eg. return 401 if un authenticated

  }

>   return {
>     statusCode: 200,
>     body: JSON.stringify({
>       message: 'Success',
>     })
>   }
> }

Reviewer: ```suggestion
  return res.json({
    message: 'Success',
  })
}
```

# Next.js

## `getServerSession`

:::warning

> This is experimental and may be removed or changed in the future.

Reviewer 2: ```suggestion
This feature is experimental and may be removed or changed in the future.
```

:::

When calling from server-side i.e. in API routes or in `getServerSideProps`, we recommend using this function instead of `getSession` to retrieve the `session` object. This method is especially useful when you are using NextAuth.js with a database. This method can _drastically_ reduce response time when used over `getSession` server-side, due to avoiding an extra `fetch` to an API Route (this is generally not recommended in Next.js). In addition, `getServerSession` will correctly update the cookie expiry time and update the session content if `callbacks.jwt` or `callbacks.session` changed something.

> Otherwise, if you only want to get the token, see [`getToken`](tutorials/securing-pages-and-api-routes#using-gettoken).

Reviewer 2: ```suggestion
Otherwise, if you only want to get the session token, see [`getToken`](tutorials/securing-pages-and-api-routes#using-gettoken).
```

> `getServerSession` requires passing the `authOptions` object. This is the same object you would pass to `NextAuth` when initializing NextAuth.js.

Reviewer 2: ```suggestion
`getServerSession` requires passing the same object you would pass to `NextAuth` when initializing NextAuth.js. To do so, you can export your NextAuth.js options in the following way:
```

In `[...nextauth.js]`:
```js 

> export const authOptions: NextAuthOptions = {

Reviewer 2: ```suggestion
import { Nextauth } from 'next-auth'
impoty type { NextAuthOptions } from 'next-auth'
 
export const authOptions: NextAuthOptions = {
```

Otherwise, if you only want to get the session token, see [`getToken`](tutorials/securing-pages-and-api-routes#using-gettoken).

`getServerSession` requires passing the same object you would pass to `NextAuth` when initializing NextAuth.js. To do so, you can export your NextAuth.js options in the following way:

In `[...nextauth.js]`:
```ts
import { NextAuth } from 'next-auth'
import type { NextAuthOptions } from 'next-auth'
 
export const authOptions: NextAuthOptions = {
  // your configs
}

export default NextAuth(authOptions);
```

In `getServerSideProps`:
```js

> import { authOptions } from 'pages/api/[...nextauth]'
> 

Reviewer: ```suggestion
import { authOptions } from 'pages/api/[...nextauth]'
import { unstable_getServerSession } from "next-auth/next"

```

export async function getServerSideProps(context) {
  const session = await getServerSession(context.req, context.res, authOptions)

  if (!session) {
    return {
      redirect: {
        destination: '/',
        permanent: false,
      },
    }
  }

  return {
    props: {
      session,
    },
  }
}
```
In API routes:
```js

> import { authOptions } from 'pages/api/[...nextauth]'

Reviewer: ```suggestion
import { authOptions } from 'pages/api/[...nextauth]'
import { unstable_getServerSession } from "next-auth/next"

```

# Next.js

> ## `getServerSession`

Reviewer 3: ```suggestion
## `unstable_getServerSession`
```

:::warning
This feature is experimental and may be removed or changed in the future.
:::

> When calling from server-side i.e. in API routes or in `getServerSideProps`, we recommend using this function instead of `getSession` to retrieve the `session` object. This method is especially useful when you are using NextAuth.js with a database. This method can _drastically_ reduce response time when used over `getSession` server-side, due to avoiding an extra `fetch` to an API Route (this is generally [not recommended in Next.js](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes)). In addition, `getServerSession` will correctly update the cookie expiry time and update the session content if `callbacks.jwt` or `callbacks.session` changed something.

Reviewer 3: ```suggestion
When calling from server-side i.e. in API routes or in `getServerSideProps`, we recommend using this function instead of `getSession` to retrieve the `session` object. This method is especially useful when you are using NextAuth.js with a database. This method can _drastically_ reduce response time when used over `getSession` server-side, due to avoiding an extra `fetch` to an API Route (this is generally [not recommended in Next.js](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes)). In addition, `unstable_getServerSession` will correctly update the cookie expiry time and update the session content if `callbacks.jwt` or `callbacks.session` changed something.
```

Otherwise, if you only want to get the session token, see [`getToken`](tutorials/securing-pages-and-api-routes#using-gettoken).

> `getServerSession` requires passing the same object you would pass to `NextAuth` when initializing NextAuth.js. To do so, you can export your NextAuth.js options in the following way:

Reviewer 3: ```suggestion
`unstable_getServerSession` requires passing the same object you would pass to `NextAuth` when initializing NextAuth.js. To do so, you can export your NextAuth.js options in the following way:
```

In `[...nextauth.js]`:
```ts
import { NextAuth } from 'next-auth'
import type { NextAuthOptions } from 'next-auth'
 
export const authOptions: NextAuthOptions = {
  // your configs
}

export default NextAuth(authOptions);
```

In `getServerSideProps`:
```js
import { authOptions } from 'pages/api/[...nextauth]'
import { unstable_getServerSession } from "next-auth/next"

export async function getServerSideProps(context) {

>   const session = await getServerSession(context.req, context.res, authOptions)

Reviewer 3: ```suggestion
  const session = await unstable_getServerSession(context.req, context.res, authOptions)
```

  if (!session) {
    return {
      redirect: {
        destination: '/',
        permanent: false,
      },
    }
  }

  return {
    props: {
      session,
    },
  }
}
```
In API routes:
```js
import { authOptions } from 'pages/api/[...nextauth]'
import { unstable_getServerSession } from "next-auth/next"

export async function handler(req, res) {

>   const session = await getServerSession(req, res, authOptions)

Reviewer 3: ```suggestion
  const session = await unstable_getServerSession(req, res, authOptions)
```
